# ==============================================================================
# FIXING COMMIT DIFF
# ==============================================================================
# File: layout/generic/SelectionMovementUtils.cpp
# Commit: e90eb44c2da3
# Full Hash: e90eb44c2da3303c038c67b7ad28b1498cd5a98c
# Author: Masayuki Nakano <masayuki@d-toybox.com>
# Date: 2024-07-19 16:21:39
# Description:
#   Bug 1897255 - Make `nsIFrame::PeekOffsetForLine` won't cross editing host boundary unless the callers allows that r=emilio
#   
#   The problems in the testcase are, `nsIFrame::PeekOffsetForLine` returns a frame
#   for any content node outside the editing host and `nsIFrame::GetLastLeaf`
#   returns a native anonymous subtree node if the frame is for the native
# ==============================================================================

diff -r bd02e7030721 -r e90eb44c2da3 layout/generic/SelectionMovementUtils.cpp
--- a/layout/generic/SelectionMovementUtils.cpp	Fri Jul 19 07:46:19 2024 +0000
+++ b/layout/generic/SelectionMovementUtils.cpp	Fri Jul 19 08:13:10 2024 +0000
@@ -26,6 +26,7 @@
 #include "nsLayoutUtils.h"
 #include "nsPresContext.h"
 #include "nsTextFrame.h"
+#include "WordMovementType.h"
 
 namespace mozilla {
 using namespace dom;
@@ -38,7 +39,7 @@
     nsIContent* aContent, uint32_t aOffset, nsDirection aDirection,
     CaretAssociationHint aHint, intl::BidiEmbeddingLevel aCaretBidiLevel,
     const nsSelectionAmount aAmount, const nsPoint& aDesiredCaretPos,
-    PeekOffsetOptions aOptions) {
+    PeekOffsetOptions aOptions, const Element* aAncestorLimiter) {
   const PrimaryFrameData frameForFocus =
       SelectionMovementUtils::GetPrimaryFrameForCaret(
           aContent, aOffset, aOptions.contains(PeekOffsetOption::Visual), aHint,
@@ -51,7 +52,7 @@
   PeekOffsetStruct pos(
       aAmount, aDirection,
       static_cast<int32_t>(frameForFocus.mOffsetInFrameContent),
-      aDesiredCaretPos, aOptions);
+      aDesiredCaretPos, aOptions, eDefaultBehavior, aAncestorLimiter);
   nsresult rv = frameForFocus.mFrame->PeekOffset(&pos);
   if (NS_FAILED(rv)) {
     return Err(rv);